src/Common/src/TypeSystem/IL/Stubs/PInvokeILEmitter.cs

```diff
@@ -175,8 +162,63 @@ public static MethodIL EmitIL(MethodDesc method, PInvokeILEmitterConfiguration p
                 return emitter.Link(method);
             }
         }
+
+        private bool IsStubRequired()
+        {
+            MethodDesc method = _methodData.TargetMethod;
+            Debug.Assert(method.IsPInvoke);
+
+            if (MarshalHelpers.UseLazyResolution(method, _methodData.ImportMetadata.Module, _methodData.PInvokeILEmitterConfiguration))
+            {
+                return true;
+            }
+            if ((_methodData.ImportMetadata.Attributes & PInvokeAttributes.SetLastError) == PInvokeAttributes.SetLastError)
+            {
+                return true;
+            }
+
+            for (int i = 0; i < _marshallers.Length; i++)
+            {
+                if (_marshallers[i].isMarshallingRequired())
+                    return true;
+            }
+            return false;
+        }
+
+        public static bool IsStubRequired(MethodDesc method, PInvokeILEmitterConfiguration pinvokeILEmitterConfiguration)
+        {
+            return new PInvokeILEmitter(method, pinvokeILEmitterConfiguration).IsStubRequired();
```

Comment:
PInvokeILEmitter is the heavy weight builder object. It should not be cached. Instead, we should be caching what it produces. I would create a subclass of ILStubMethodIL with the extra bit to cache whether there is any marshalling.

There will be a bit of wasted effort in generating the IL when the marshalling is not required, but I do not think it is a problem. The nice side-effect is that there will be much less duplication in the algorithm that computes whether marshalling is required.